src/OpenTK.Compute/OpenCL/CL.cs

```diff
@@ -1,1543 +1,3192 @@
+        /// </summary>
+        public static CLResultCode CompileProgram(
+           this CLProgram program,
+            CLDevice[] deviceList,
+            string options,
+            CLProgram[] inputHeaders,
+            string[] headerIncludeNames,
+            ClEventCallback callback,
+            IntPtr userData)
+        {
+            return CompileProgram(program, (uint)(deviceList?.Length ?? 0), deviceList, options, (uint)(inputHeaders?.Length ?? 0), inputHeaders,
+               headerIncludeNames, callback == null ? IntPtr.Zero : Marshal.GetFunctionPointerForDelegate(callback), userData);
+        }
+
+        /// <summary>
+        /// Introduced in OpenCL 1.2
+        /// </summary>
+        [DllImport(LibName, CallingConvention = CallingConvention, EntryPoint = "clLinkProgram")]
+        public static extern CLProgram LinkProgram(
+            [In] this CLContext context,
+            [In] uint numberOfDevices,
+            [In] CLDevice[] deviceList,
+            [In] string options,
+            [In] uint numberOfPrograms,
+            [In] CLProgram[] inputPrograms,
+            [In] IntPtr notificationCallback,
+            [In] IntPtr userData,
+            [Out] out CLResultCode resultCode);
+
+        /// <summary>
+        /// Introduced in OpenCL 1.2
+        /// </summary>
+        public static CLProgram LinkProgram(
+            this CLContext context,
+            CLDevice[] deviceList,
+            string options,
+            CLProgram[] inputPrograms,
+            ClEventCallback callback,
+            IntPtr userData,
+            out CLResultCode resultCode)
+        {
+            return LinkProgram(context, (uint)(deviceList?.Length ?? 0), deviceList, options, (uint)(inputPrograms?.Length ?? 0), inputPrograms,
+               callback == null ? IntPtr.Zero : Marshal.GetFunctionPointerForDelegate(callback), userData, out resultCode);
+        }
+
+        /// <summary>
+        /// Introduced in OpenCL 2.2
+        /// </summary>
+        [DllImport(LibName, CallingConvention = CallingConvention, EntryPoint = "clSetProgramSpecializationConstant")]
+        public static extern CLResultCode SetProgramSpecializationConstant(
+            [In] this CLProgram program,
+            [In] uint specId,
+            [In] nuint specSize,
+            [In] IntPtr specValue);
+
+        /// <summary>
+        /// Introduced in OpenCL 2.2
+        /// </summary>
+        public static unsafe CLResultCode SetProgramSpecializationConstant<T>(
+            this CLProgram program,
+            uint specId,
+            T specValue)
+            where T : unmanaged
+        {
+            if (specValue is bool)
+            {
+                ushort shortVal = (bool)Convert.ChangeType(specValue, typeof(bool)) ? (ushort)1 : (ushort)0;
+                ushort* shortArg = &shortVal;
+                return SetProgramSpecializationConstant(program, specId, (nuint)sizeof(ushort), (IntPtr)shortArg);
+            }
```

Comment:
This should really be a separate overload for bool and not a special case here (if we are even going to have a bool overload. Does OpenCL have a bool type?).
And it should be using pattern matching (`if (specValue is bool specBool)`) if this is to stay.

Reply:
In the documentation, it states that booleans should have a size of 1. I've updated the code to better fit this

`spec_size specifies the size in bytes of the data pointed to by spec_value. This should be 1 for boolean constants. For all other constant types this should match the size of the specialization constant in the module.`

https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_setting_spir_v_specialization_constants